pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,197 @@
+	providerConfigPath = specPath.Child("providerConfig")
+	nwPath             = specPath.Child("networking")
+	providerPath       = specPath.Child("provider")
+	infraConfigPath    = providerPath.Child("infrastructureConfig")
+	cpConfigPath       = providerPath.Child("ControlPlaneConfig")
+	workersPath        = providerPath.Child("workers")
+)
+
+func (v *Shoot) validateShootCreation(ctx context.Context, shoot *core.Shoot) error {
+	valContext, err := newValidationContext(v.decoder, shoot)
+	if err != nil {
+		return err
+	}
+
+	if err := v.validateInfraAgainstCloudProfile(ctx, shoot, valContext.infraConfig, infraConfigPath); err != nil {
+		return err
+	}
+	if err := v.validateCpAgainstCloudProfile(ctx, shoot, valContext.cpConfig, cpConfigPath); err != nil {
+		return err
+	}
+
+	return v.validateShoot(ctx, valContext)
+}
+
+func (v *Shoot) validateShootUpdate(ctx context.Context, oldShoot, shoot *core.Shoot) error {
+	oldValContext, err := newValidationContext(v.decoder, oldShoot)
+	if err != nil {
+		return err
+	}
+
+	valContext, err := newValidationContext(v.decoder, shoot)
+	if err != nil {
+		return err
+	}
+
+	if errList := openstackvalidation.ValidateInfrastructureConfigUpdate(oldValContext.infraConfig, valContext.infraConfig, infraConfigPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Only validate against cloud profile when related configuration is updated.
+	// This ensures that already running shoots won't break after constraints were removed from the cloud profile.
+	if oldValContext.infraConfig.FloatingPoolName != valContext.infraConfig.FloatingPoolName {
+		if err := v.validateInfraAgainstCloudProfile(ctx, shoot, valContext.infraConfig, infraConfigPath); err != nil {
+			return err
+		}
+	}
+
+	var (
+		oldCpConfig = oldValContext.cpConfig
+		cpConfig    = valContext.cpConfig
+	)
+	if errList := openstackvalidation.ValidateControlPlaneConfigUpdate(oldCpConfig, cpConfig, cpConfigPath); len(errList) != 0 {
+		return errList.ToAggregate()
+	}
+
+	// Only validate against cloud profile when related configuration is updated.
+	// This ensures that already running shoots won't break after constraints were removed from the cloud profile.
+	if oldCpConfig.LoadBalancerProvider != cpConfig.LoadBalancerProvider ||
+		oldCpConfig.Zone != cpConfig.Zone ||
+		!equality.Semantic.DeepEqual(oldCpConfig.LoadBalancerClasses, cpConfig.LoadBalancerClasses) {
+		if err := v.validateCpAgainstCloudProfile(ctx, shoot, cpConfig, cpConfigPath); err != nil {
+			return err
+		}
+	}
+
+	if errList := openstackvalidation.ValidateWorkersUpdate(oldShoot.Spec.Provider.Workers, shoot.Spec.Provider.Workers, workersPath); len(errList) > 0 {
+		return errList.ToAggregate()
+	}
+
+	return v.validateShoot(ctx, valContext)
+}
+
+func (v *Shoot) validateShoot(ctx context.Context, context *validationContext) error {
+	if errList := openstackvalidation.ValidateNetworking(context.shoot.Spec.Networking, nwPath); len(errList) != 0 {
```

Comment:
Is there a specific reason, why you don't collect the errors from the different validation functions to return a combined error list?
If not I would suggest appending the returned errors to a `field.ErrorList` and returning it at the end of the function :)